### PR TITLE
[java] Adding addExternalInitializers and addInitializer to OrtSession.SessionOptions

### PR DESCRIPTION
**Pull request by Craigacp**
### Description
Adds support for adding external initializers or overriding initializers to a session options from Java.

### Motivation and Context
We want to instantiate large models from Java without filesystem access.

cc @yuslepukhin

### CODE REVIEW

File: java/src/main/native/ai_onnxruntime_OrtSession_SessionOptions.c [hunk 1]
```diff
@@ -350,6 +350,85 @@ JNIEXPORT void JNICALL Java_ai_onnxruntime_OrtSession_00024SessionOptions_addFre
...
+  const OrtApi* api = (const OrtApi*)apiHandle;
+  OrtSessionOptions* options = (OrtSessionOptions*) optionsHandle;
+
+  // Extract the string chars
+  const char* cName = (*jniEnv)->GetStringUTFChars(jniEnv, name, NULL);
+
+  // Cast the onnx value
+  const OrtValue* tensor = (const OrtValue*) tensorHandle;
+
+  checkOrtStatus(jniEnv,api,api->AddInitializer(options,cName,tensor));
```
Comment (RyanUnderhill) [inline java/src/main/native/ai_onnxruntime_OrtSession_SessionOptions.c:426]: This function throws on errors? Won't that skip over the ReleaseStringUTFChars?

File: java/src/main/native/ai_onnxruntime_OrtSession_SessionOptions.c [hunk 2]
```diff
@@ -350,6 +350,85 @@ JNIEXPORT void JNICALL Java_ai_onnxruntime_OrtSession_00024SessionOptions_addFre
...
+  jlong* initializersArr = (*jniEnv)->GetLongArrayElements(jniEnv, handlesArray, NULL);
+
+  for (size_t i = 0; i < arrLength; i++) {
+    // Extract the string chars and cast the tensor
+    javaNameStrings[i] = (*jniEnv)->GetObjectArrayElement(jniEnv, namesArray, (jint) i);
+    names[i] = (*jniEnv)->GetStringUTFChars(jniEnv, javaNameStrings[i], NULL);
+    initializers[i] = (const OrtValue*) initializersArr[i];
+  }
+
+  checkOrtStatus(jniEnv,api,api->AddExternalInitializers(options,names,initializers,arrLength));
```
Comment (RyanUnderhill) [inline java/src/main/native/ai_onnxruntime_OrtSession_SessionOptions.c:394]: This function throws on errors? Won't that skip over the cleanup below?

File: java/src/main/java/ai/onnxruntime/OrtSession.java [hunk 3]
```diff
@@ -785,6 +785,53 @@ public Map<String, String> getConfigEntries() {
...
+     * {@link #addInitializer(String, OnnxTensorLike)}. The initializers must be created from 
+     * {@link java.nio.Buffer} objects.
+     *
+     * @param initializers The map of names to initializers.
+     * @throws OrtException If the initializers could not be loaded.
+     */
+    public void addExternalInitializers(Map<String, OnnxTensorLike> initializers)
+        throws OrtException {
+      checkClosed();
+      if (!initializers.isEmpty()) {
```
Comment (RyanUnderhill) [inline java/src/main/java/ai/onnxruntime/OrtSession.java:803]: Minor nit, but could invert the logic to make it early exit vs having the rest of the function in the if{} block?

File: java/src/main/native/ai_onnxruntime_OrtSession_SessionOptions.c (hunk 2 above)
Comment (Craigacp) [reply java/src/main/native/ai_onnxruntime_OrtSession_SessionOptions.c:394]: It throws an error up into the JVM, but the rest of the native function completes before control is returned to the Java code which propagates the exception. So the cleanup happens before any more Java code executes.

File: java/src/main/native/ai_onnxruntime_OrtSession_SessionOptions.c (hunk 1 above)
Comment (Craigacp) [reply java/src/main/native/ai_onnxruntime_OrtSession_SessionOptions.c:426]: Same comment as above, the checkOrtStatus method throws a Java exception, but the JNI code completes before it is propagated.

File: java/src/main/java/ai/onnxruntime/OrtSession.java (hunk 3 above)
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OrtSession.java:803]: Like this:
```java
      checkClosed();
      if (initializers.isEmpty()) {
        return;
      }
      String[] names = new String[initializers.size()];
      long[] handles = new long[initializers.size()];
      int i = 0;
      for (Map.Entry<String, OnnxTensorLike> e : initializers.entrySet()) {
        names[i] = e.getKey();
        handles[i] = e.getValue().nativeHandle;
        i++;
      }
      addExternalInitializers(OnnxRuntime.ortApiHandle, nativeHandle, names, handles);
``` ?

Comment (RyanUnderhill) [reply java/src/main/java/ai/onnxruntime/OrtSession.java:803]: Yep, up to you, just something I've found cleaner and avoids indenting deeper as the function goes, and makes it clear there's no 'else'.

File: java/src/main/native/ai_onnxruntime_OrtSession_SessionOptions.c (hunk 2 above)
Comment (RyanUnderhill) [reply java/src/main/native/ai_onnxruntime_OrtSession_SessionOptions.c:394]: Ah, ok. I was wondering how C was throwing exceptions, that makes sense. It's a Java thing. All good.

File: java/src/main/java/ai/onnxruntime/OrtSession.java (hunk 3 above)
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OrtSession.java]: Done.

**Review by RyanUnderhill — DISMISSED** [review]

**Comment by RyanUnderhill** [comment]
Look like source formatting is not passing:

```
* What went wrong:
Execution failed for task ':spotlessJavaCheck'.
> The following files had format violations:
```

**Comment by Craigacp** [comment]
Fixed.

**Review by RyanUnderhill — APPROVED** [review]

**Pull request merged**